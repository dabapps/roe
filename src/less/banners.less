.banner {
  position: fixed;
  width: 100%;
  bottom: 0;
  left: 0;
  top: auto;
  height: @banner-height;
  background-color: @banner-background;
  border-top: @banner-border;
  z-index: 1100;
  transform: translate(0, 100%);
  transition: transform 0.3s ease-in-out;

  &.top {
    bottom: auto;
    top: 0;
    transform: translate(0, -100%);
    border-top: @border-none;
    border-bottom: @banner-border;
  }

  &.open {
    transform: translate(0, 0);
  }
}

.cookie-banner {
  z-index: 1200;
}
