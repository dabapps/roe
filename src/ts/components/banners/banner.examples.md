#### Example

```js static
<Banner
  open
  position="top"
>
  <Container>
    <Row>
      <Column xs={10}>
        <p>Roe Banner</p>
      </Column>
      <Column xs={2}>
        <Button
          className={'margin-top-base float-right'}
        >
          Click
      </Button>
      </Column>
    </Row>
  </Container>
</Banner>
```

#### Less variables

```less
@banner-background: @body-background;
@banner-border: @border-base;
@banner-height: auto;
```
