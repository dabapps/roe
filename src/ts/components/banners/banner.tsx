import * as classNames from 'classnames';
import * as React from 'react';

import { memoWithComponentProp } from '../../utils';
import {
  FunctionComponentOptionalComponentProp,
  OptionalComponentProp,
} from '../../types';

export interface BannerProps {
  /**
   * If set, displays the component, otherwise it is hidden
   * @default true
   */
  open?: boolean;
  /**
   * Positions the element at the 'top' or 'bottom' of the screen
   * @default 'bottom'
   */
  position?: 'top' | 'bottom';
}

/**
 * A Banner component that displays fixed to the top or bottom of the screen.
 */
const Banner: FunctionComponentOptionalComponentProp<'div', BannerProps> = (
  props: OptionalComponentProp<'div'> & BannerProps
): React.ReactElement => {
  const {
    className,
    children,
    open = true,
    position = 'bottom',
    component: Component = 'div',
    ...remainingProps
  } = props;

  return (
    <Component
      {...remainingProps}
      className={classNames('banner', open && 'open', position, className)}
    >
      {children}
    </Component>
  );
};

export default memoWithComponentProp(Banner);
