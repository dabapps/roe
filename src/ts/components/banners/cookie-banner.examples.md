#### Example

```js static
<CookieBanner
  position="top"
  render={({ dismiss }) => (
    <Container>
      <Row>
        <Column xs={10}>
          <p>We use cookies! Roe is awesome</p>
        </Column>
        <Column xs={2}>
          <Button onClick={dismiss} className="margin-top-base float-right">
            Accept
          </Button>
        </Column>
      </Row>
    </Container>
  )}
/>
```

#### Less variables

```css
@banner-background: @body-background;
@banner-border: @border-base;
@banner-height: auto;
```
