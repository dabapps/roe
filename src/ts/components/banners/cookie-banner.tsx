import * as classNames from 'classnames';
import * as cookie from 'cookie';
import * as React from 'react';
import { HTMLProps, PureComponent } from 'react';
import { ComponentProps } from '../../types';
import Banner from './banner';

export type Render = (
  props: { dismiss: () => void }
) => React.ReactElement<any>;

export interface CookieBannerProps
  extends ComponentProps,
    HTMLProps<HTMLElement> {
  /**
   * Takes a component as a function and renders as a child
   */
  render: Render;
  /**
   * Positions the element at the 'top' or 'bottom' of the screen
   * @default 'bottom'
   */
  position?: 'top' | 'bottom';
}

export interface CookieBannerState {
  dismissed: boolean;
}

/**
 * A [Banner](#banner) component that is permanently dismissed after setting a cookie.
 * This component takes a render prop, which can be a component or function, that is passed a dismiss prop
 * which you can then apply as an onClick prop to an element of your choice.
 */
export class CookieBanner extends PureComponent<
  CookieBannerProps,
  CookieBannerState
> {
  public constructor(props: CookieBannerProps) {
    super(props);

    this.state = {
      dismissed: Boolean(cookie.parse(document.cookie)['cookies-accepted']),
    };
  }

  public render() {
    const {
      ref,
      className,
      children,
      render,
      position = 'bottom',
      ...remainingProps
    } = this.props;

    const { dismissed } = this.state;

    return (
      <Banner
        {...remainingProps}
        position={position}
        open={!dismissed}
        className={classNames('cookie-banner', className)}
      >
        {render && render({ dismiss: this.setCookie })}
      </Banner>
    );
  }

  private setCookie = () => {
    document.cookie = cookie.serialize('cookies-accepted', 'true');
    this.setState({
      dismissed: true,
    });
  };
}

export default CookieBanner;
