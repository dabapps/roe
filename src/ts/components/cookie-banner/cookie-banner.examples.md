#### Example

```js
class CookieBannerExample extends React.Component {

  render () {

    retrun (
      <CookieBanner
        render={({dismiss}) => (
          <Container>
            <Row>
              <Column xs={10}>
                <p>Roe Cookie Banner</p>
              </Column>
              <Column xs={2}>
                <Button
                  onClick={dismiss}
                  className={'margin-top-base float-right'}
                >
                  Accept
                </Button>
              </Column>
            </Row>
          </Container>
        )}
      />
    )
  }
}

<CookieBannerExample />
```
