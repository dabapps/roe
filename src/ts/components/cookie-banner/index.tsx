import * as classNames from 'classnames';
import * as React from 'react';
import * as cookie from 'cookie';
import { HTMLProps, PureComponent } from 'react';
import { ComponentProps } from '../../types';
import Banner from  '../banner';

type IRender = (props: {dismiss: () => void}) => React.ReactElement<any>;

interface CookieBannerProps extends ComponentProps, HTMLProps<HTMLElement> {
  render: IRender;
  top?: boolean;
}

interface CookieBannerState {
  dismissed: boolean;
}

class CookieBanner extends PureComponent<CookieBannerProps, CookieBannerState> {

  public constructor(props: CookieBannerProps) {
    super(props);

    this.state = {
      dismissed: Boolean(cookie.parse(document.cookie)['cookies-accepted'])
    };
  }

  public render () {
    const {
      ref,
      className,
      children,
      render,
      top = false,
      ...remainingProps
    } = this.props;

    const { dismissed } = this.state;

    return (
      <Banner
        {...remainingProps}
        top={top}
        open={!dismissed}
        className={classNames('cookie-banner', className)}
      >
        {render && render({dismiss: this.setCookie})}
      </Banner>
    )
  }

  private setCookie = () =>  {
    document.cookie = cookie.serialize('cookies-accepted', 'true');
    this.setState({
      dismissed: true
    });
  }
}

export default CookieBanner;
