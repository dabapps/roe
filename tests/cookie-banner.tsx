import * as enzyme from 'enzyme';
import * as React from 'react';
import * as renderer from 'react-test-renderer';

import {
  Button,
  Column,
  Container,
  Row
} from '../src/ts';

import { CookieBanner } from '../src/ts/components/banners/cookie-banner';

describe('CookieBanner', () => {

  it('should match snapshot and take required render prop', () => {
    const tree = renderer.create(
      <CookieBanner
        render={({ dismiss }) => (
          <Container>
            <Row>
              <Column xs={10}>
                <p>We use cookies! Roe is awesome</p>
              </Column>
              <Column xs={2}>
                <Button
                  onClick={dismiss}
                  className={'margin-top-base float-right'}
                >
                  Accept
                </Button>
              </Column>
            </Row>
          </Container>
        )}
      />
    );

    expect(tree).toMatchSnapshot();
  });

  it('should take regular element attributes', () => {
    const tree = renderer.create(
      <CookieBanner
        className="my-class"
        render={({ dismiss }) => (
          <Container>
            <Row>
              <Column xs={10}>
                <p>We use cookies! Roe is awesome</p>
              </Column>
              <Column xs={2}>
                <Button
                  onClick={dismiss}
                  className={'margin-top-base float-right'}
                >
                  Accept
                </Button>
              </Column>
            </Row>
          </Container>
        )}
      />
    );

    expect(tree).toMatchSnapshot();
  });

  it('should take an optional position prop', () => {
    const tree = renderer.create(
      <CookieBanner
        position={'top'}
        render={({ dismiss }) => (
          <Container>
            <Row>
              <Column xs={10}>
                <p>We use cookies! Roe is awesome</p>
              </Column>
              <Column xs={2}>
                <Button
                  onClick={dismiss}
                  className={'margin-top-base float-right'}
                >
                  Accept
                </Button>
              </Column>
            </Row>
          </Container>
        )}
      />
    );

    expect(tree).toMatchSnapshot();
  });

  it('should hide the banner on click', () => {

    const instance = enzyme.shallow(
      <CookieBanner
        render={({ dismiss }) => (
          <Button
            onClick={dismiss}
            className={'button'}
          >
            Accept
          </Button>
        )}
      />
    );

    expect(instance).toMatchSnapshot();

    instance.find('.button').simulate('click');
    expect(instance).toMatchSnapshot();
  });

});
