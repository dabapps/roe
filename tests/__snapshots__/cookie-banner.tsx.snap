// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CookieBanner should hide the banner on click 1`] = `
<Memo(CookieBanner)
  render={[Function]}
>
  <Memo(Banner)
    className="cookie-banner"
    component="div"
    open={true}
    position="bottom"
  >
    <div
      className="banner open bottom cookie-banner"
    >
      <Memo(Container)>
        <div
          className="container"
        >
          <Memo(Row)>
            <div
              className="row"
            >
              <Memo(Column)
                xs={10}
              >
                <div
                  className="column xs-10"
                >
                  <p>
                    We use cookies! Roe is awesome
                  </p>
                </div>
              </Memo(Column)>
              <Memo(Column)
                xs={2}
              >
                <div
                  className="column xs-2"
                >
                  <Memo(Button)
                    className="margin-top-base float-right"
                    onClick={[Function]}
                  >
                    <button
                      className="button margin-top-base float-right"
                      onClick={[Function]}
                    >
                      Accept
                    </button>
                  </Memo(Button)>
                </div>
              </Memo(Column)>
            </div>
          </Memo(Row)>
        </div>
      </Memo(Container)>
    </div>
  </Memo(Banner)>
</Memo(CookieBanner)>
`;

exports[`CookieBanner should hide the banner on click 2`] = `
<Memo(CookieBanner)
  render={[Function]}
>
  <Memo(Banner)
    className="cookie-banner"
    component="div"
    open={false}
    position="bottom"
  >
    <div
      className="banner bottom cookie-banner"
    >
      <Memo(Container)>
        <div
          className="container"
        >
          <Memo(Row)>
            <div
              className="row"
            >
              <Memo(Column)
                xs={10}
              >
                <div
                  className="column xs-10"
                >
                  <p>
                    We use cookies! Roe is awesome
                  </p>
                </div>
              </Memo(Column)>
              <Memo(Column)
                xs={2}
              >
                <div
                  className="column xs-2"
                >
                  <Memo(Button)
                    className="margin-top-base float-right"
                    onClick={[Function]}
                  >
                    <button
                      className="button margin-top-base float-right"
                      onClick={[Function]}
                    >
                      Accept
                    </button>
                  </Memo(Button)>
                </div>
              </Memo(Column)>
            </div>
          </Memo(Row)>
        </div>
      </Memo(Container)>
    </div>
  </Memo(Banner)>
</Memo(CookieBanner)>
`;

exports[`CookieBanner should match snapshot and take required render prop 1`] = `
<div
  className="banner open bottom cookie-banner"
>
  <div
    className="container"
  >
    <div
      className="row"
    >
      <div
        className="column xs-10"
      >
        <p>
          We use cookies! Roe is awesome
        </p>
      </div>
      <div
        className="column xs-2"
      >
        <button
          className="button margin-top-base float-right"
          onClick={[Function]}
        >
          Accept
        </button>
      </div>
    </div>
  </div>
</div>
`;

exports[`CookieBanner should take an optional position prop 1`] = `
<div
  className="banner open top cookie-banner"
>
  <div
    className="container"
  >
    <div
      className="row"
    >
      <div
        className="column xs-10"
      >
        <p>
          We use cookies! Roe is awesome
        </p>
      </div>
      <div
        className="column xs-2"
      >
        <button
          className="button margin-top-base float-right"
          onClick={[Function]}
        >
          Accept
        </button>
      </div>
    </div>
  </div>
</div>
`;

exports[`CookieBanner should take regular element attributes 1`] = `
<div
  className="banner open bottom cookie-banner my-class"
>
  <div
    className="container"
  >
    <div
      className="row"
    >
      <div
        className="column xs-10"
      >
        <p>
          We use cookies! Roe is awesome
        </p>
      </div>
      <div
        className="column xs-2"
      >
        <button
          className="button margin-top-base float-right"
          onClick={[Function]}
        >
          Accept
        </button>
      </div>
    </div>
  </div>
</div>
`;
